source/secure_tunneling.c

```diff
@@ -11,38 +11,55 @@
 #define INVALID_STREAM_ID 0
 #define MAX_ST_PAYLOAD 64512
 #define PAYLOAD_BYTE_LENGTH_PREFIX 2
+#define PING_TASK_INTERVAL ((uint64_t)20 * 1000000000)
 
 #define UNUSED(x) (void)(x)
 
-static void s_send_websocket_ping(struct aws_secure_tunnel *secure_tunnel) {
-    if (!secure_tunnel->websocket) {
+static void s_send_websocket_ping(struct aws_websocket *websocket) {
+    if (!websocket) {
         return;
     }
 
     struct aws_websocket_send_frame_options frame_options;
     AWS_ZERO_STRUCT(frame_options);
     frame_options.opcode = AWS_WEBSOCKET_OPCODE_PING;
     frame_options.fin = true;
-    aws_websocket_send_frame(secure_tunnel->websocket, &frame_options);
+    aws_websocket_send_frame(websocket, &frame_options);
 }
 
+struct ping_task_context {
+    struct aws_allocator *allocator;
+    struct aws_event_loop *event_loop;
+
+    struct aws_atomic_var task_cancelled;
+    struct aws_websocket *websocket;
+};
+
 static void s_ping_task(struct aws_task *task, void *user_data, enum aws_task_status task_status) {
     AWS_LOGF_TRACE(AWS_LS_IOTDEVICE_SECURE_TUNNELING, "s_ping_task");
 
-    if (task_status != AWS_TASK_STATUS_RUN_READY) {
-        AWS_LOGF_ERROR(AWS_LS_IOTDEVICE_SECURE_TUNNELING, "task_status is not ready. Do nothing.");
+    struct ping_task_context *ping_task_context = user_data;
+
+    if (task_status == AWS_TASK_STATUS_CANCELED) {
+        AWS_LOGF_INFO(
+            AWS_LS_IOTDEVICE_SECURE_TUNNELING, "task_status is AWS_TASK_STATUS_CANCELED. Cleaning up ping task.");
+        aws_mem_release(ping_task_context->allocator, ping_task_context);
         return;
     }
 
-    struct aws_secure_tunnel *secure_tunnel = user_data;
-    s_send_websocket_ping(secure_tunnel);
+    const size_t task_cancelled = aws_atomic_load_int(&ping_task_context->task_cancelled);
+    if (task_cancelled) {
+        AWS_LOGF_INFO(AWS_LS_IOTDEVICE_SECURE_TUNNELING, "task_cancelled is true. Cleaning up ping task.");
+        aws_mem_release(ping_task_context->allocator, ping_task_context);
+        return;
+    }
 
-    // Schedule the next task
-    struct aws_event_loop *event_loop =
-        aws_event_loop_group_get_next_loop(secure_tunnel->config.bootstrap->event_loop_group);
+    s_send_websocket_ping(ping_task_context->websocket);
+
+    /* Schedule the next task */
     uint64_t now;
-    aws_event_loop_current_clock_time(event_loop, &now);
-    aws_event_loop_schedule_task_future(event_loop, task, now + (uint64_t)20 * 1000000000);
+    aws_event_loop_current_clock_time(ping_task_context->event_loop, &now);
```

Comment:
I think we still have a potential race condition where shutdown/external cleanup happens as we reach this point.  The ping context is safe since it's on the heap and the task function is solely responsible for cleaning it up, but the ping task itself is not safe since it's an embedded structure in the secure tunnel so it someone cleans up the secure tunnel right after connection shutdown and the ping is mid-execution and so gets rescheduled, we will end up scheduling a task that is freed memory.  

I think you could fix it by moving the task structure into the ping task context.